main.go

```diff
@@ -142,19 +142,17 @@ func (s *prometheusServer) Start(ctx context.Context) error {
 		return err
 	}
 
+	subCtx, cancel := context.WithCancel(ctx)
 	go func() {
-		// TODO: change context, so it can be closed when
-		// this function returns an error
-		if err := certWatcher.Start(ctx); err != nil {
+		if err := certWatcher.Start(subCtx); err != nil {
 			setupLog.Error(err, "certificate watcher error")
 		}
+		cancel()
 	}()
 
 	idleConnsClosed := make(chan struct{})
 	go func() {
-		// TODO: make sure that the goroutine finishes when
-		// this function returns an error
-		<-ctx.Done()
+		<-subCtx.Done()
 		setupLog.Info("shutting down Prometheus metrics server")
 
 		if err := server.Shutdown(context.Background()); err != nil {
```

Comment:
Please, add a check here, so that error is not logged if the server is already shutdown.
```
!errors.Is(err, http.ErrServerClosed)
```